packages/worker/src/sinks.ts

```diff
@@ -17,18 +17,76 @@
+ * individually. Consequently, sink functions that internally perform async operations may end up
+ * executing in parallel. If strict ordering of some sink function execution is required, consider
+ * enabling the `callSerially` option on these functions. Note however that serializing sink
+ * function calls may increase total execution time, and therefore increase the risk of Workflow
+ * Task Timeout.
+ *
+ * Please note that sink functions only provide best-effort delivery semantics, which is generally
+ * suitable for log messages and general metrics collection. However, in various situations, a sink
+ * function call may execute more than once even though the sink function is configured with
+ * `callInReplay: false`. Similarly, sink function execution errors only results in log messages,
+ * and are therefore likely to go unnoticed. For use cases that require _at-least-once_ execution
+ * guarantees, please consider using local activities instead. For use cases that require
+ * _exactly-once_ or _at-most-once_ execution guarantees, please consider using regular activities.
  */
 export interface InjectedSinkFunction<F extends SinkFunction> {
+  /**
+   * The implementation function for sink function `F`
+   */
   fn(info: WorkflowInfo, ...args: Parameters<F>): void | Promise<void>;
+
+  /**
+   * Whether or not the function will be called during Workflow replay.
+   *
+   * Take note that setting `callDuringReplay` to `false` (or leaving it unset) doesn't guarantee
+   * that the sink function will only ever run once for a particular Workflow execution at a
+   * particular point of its history. In particular, calls to sink functions will be executed
+   * even if the current workflow task ends up failling or timing out. In such situations, a call to
+   * a sink function configured with `callDuringReplay: false` will be executed again, since
+   * the workflow task is not being replayed (ie. retrying a workflow task is not the same as
+   * replaying it).
+   *
+   * For use cases that require _at-most-once_ or _exactly-once_ guarantees, please consider using
+   * a regular activity instead.
+   *
+   * Defaults to `false`.
+   */
   callDuringReplay?: boolean;
+
+  /**
+   * If set to `true`, execution calls to this sink function will be serialized, ensuring strict
+   * ordering of this sink function execution with respect to any other sink function.
+   *
+   * More formally, any pending executions of sink functions prior to this one will be awaited for
+   * completion before this one is executed, and completion of this sink function call will be
+   * awaited for completion before any subsequent sink function call is executed.
+   *
+   * Note that serializing sink function calls may increase total execution time, and therefore
+   * increase the risk of Workflow Task Timeout.
+   *
+   * Defaults to `false`.
+   */
+  callSerially?: boolean;
```

Comment:
Instead of a boolean make it a `concurrency: SinkConcurrency` enum, that we can (in the future) choose to run all sinks for an interface serially vs. sinks of the same function.

Reply:
Also would be nice to allow failing a WFT on sink failure, but that can also wait for another PR.